packages/remix-react/components.tsx

```diff
@@ -594,6 +594,65 @@ export type ScriptProps = Omit<
   | "suppressHydrationWarning"
 >;
 
+function remixContextJsString(handoff: string) {
+  return `window.__remixContext = ${handoff};`;
+}
+
+function remixEntryJsString(
+  manifest: AssetsManifest,
+  matches: AgnosticDataRouteMatch[],
+  handoff?: string
+) {
+  let { routes } = manifest;
+  return [
+    manifest.hmr?.runtime
+      ? `import ${JSON.stringify(manifest.hmr.runtime)};`
+      : "",
+    `import ${JSON.stringify(manifest.url)};`,
+    ...matches.map(
+      ({ route: { id } }, i) =>
+        `import * as route${i} from ${JSON.stringify(routes[id].module)};`
+    ),
+    // If we got a handoff, insert it after the imports (used for SPA mode library mode)
+    handoff ? remixContextJsString(handoff) : "",
+    `window.__remixRouteModules = {${matches
+      .map((match, index) => `${JSON.stringify(match.route.id)}:route${index}`)
+      .join(",")}}`,
+    `import(${JSON.stringify(manifest.entry.module)});`,
+  ]
+    .filter((l) => l)
+    .join("\n");
+}
+
+// This would be a publicly exported utility function that folks could use in
+// their `entry.server.tsx` if they wanted to run Remix SPA Mode as a "Library"
+// where the output should be a JS file, not an HTML file.  This allows for the
+// Remix app to be embedded in HTML not controlled by Remix.  The vite plugin
+// will check the Content-Type on the response from entry.server and write out
+// build/client/assets.index.js if "application/javascript" is returned.
+//
+// When used in this manner:
+//  - `HydrateFallback` _must_ be provided in root, the default implementation
+//     renders a complete document
+//  - `HydrateFallBack` should match the content of `<div id="app">` exactly
+//  - `HydrateFallBack` should not render `<Scripts>`
+//
+// Example entry.server.tsx:
+//   export default function handleRequest(_, __, ___, remixContext) {
+//     let js = UNSAFE_remixSpaModeLibraryJs(remixContext);
+//     return new Response(js, {
+//       headers: { "Content-Type": "application/javascript" },
+//       status: 200,
+//     });
+//   }
+export function UNSAFE_remixSpaModeLibraryJs(ctx: EntryContext) {
+  return remixEntryJsString(
+    ctx.manifest,
+    ctx.staticHandlerContext.matches,
+    ctx.serverHandoffString
+  );
+}
```

Comment:
This is the key - right now, the suggestion in the RFC is to basically recreate this logic which is all implementation details.  If we expose a utility to do this outside of a React render tree, it can be used without rendering `<RemixServer>` in `entry-server.tsx`

Reply:
Thanks so much for looking into this Matt. Exporting this logic as a utility will make the approach much simpler.

With this, teams can incrementally migrate legacy PHP/ASP.NET sites to Remix by embedding a Remix SPA.